src/main/java/com/google/sps/data/Journey.java

```diff
@@ -0,0 +1,180 @@
+package com.google.sps.data;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Date;
+import java.util.Map;
+import java.lang.reflect.Array;
+import com.google.maps.errors.ApiException;
+import java.io.IOException;
+import java.lang.InterruptedException;
+
+import java.util.HashMap;
+
+/**
+ * Class that represents a delivery journey.
+ */
+public class Journey {
+  private int minTime;
+  private int numberOfWaypoints;
+  private ArrayList<Point> orderedWaypoints;
+  // Variable pathFinder will be used for getting the path between two Point objects.
+  // private PathFinder pathFinder;
+  private ArrayList<Pair> restrictions;
+  private Point start;
+  private HashMap<Point, Integer> waypoints;
+
+  /**
+   * Helper class used for computing the bestTime matrix and order restrictions between pairs of 
+   * waypoints.
+   */
+  private class Pair {
+    public int first;
+    public int second;
+    public Pair(int first, int second) {
+      this.first = first;
+      this.second = second;
+    }
+  }
+
+  /**
+   *  Initializes a journey object that starts at start and computes paths using pathFinder.
+   */
+  public Journey(Point start /** , PathFinder pathFinder*/) {
+    this.start = start;
+    // TODO[ak47na]: add PathFinder interface and the classes that implement it: 
+    // GoogleMapsPathFinder used for production and ManhattanDistancePathFinder for testing.
+    // this.pathFinder = pathFinder;
+    waypoints = new HashMap<>();
+    // Hashmap with keys representing Point objects and values representing their index. E.g. when
+    // a new Point is added to the hashmap, its value will be the equal to the number of elements.
+    // The hashmap is preferred to an ArrayList so that Point objects can be removed efficiently.
+    
+    restrictions = new ArrayList<>();
+    // Array of pairs (a, b) meaning that waypoint with value a in waypoints must be visited before
+    // waypoint b.
+    
+    minTime = 0;
+    numberOfWaypoints = 0;
+  }
+
+  public int getNumberOfWaypoints() {
+    return numberOfWaypoints;
+  }
+  public ArrayList<Pair> getRestrictions() {
+    return restrictions;
+  }
+  public HashMap<Point, Integer> getWaypoints() {
+    return waypoints;
+  }
+
+  /** 
+   * Returns the waypoints of the journey as an ArrayList so that they can be accessed by their 
+   * value in the waypoints hashmap, which is the index in the ArrayList.
+   */
+  private ArrayList<Point> getWaypointsList() {
+    ArrayList<Point> waypointsList = new ArrayList<>();
+    waypointsList.addAll(Collections.nCopies(numberOfWaypoints, start));
+    for (Map.Entry<Point, Integer> entry : waypoints.entrySet()) {
+      waypointsList.set(entry.getValue(), entry.getKey());
+    }
+    return waypointsList;
+  }
+
+  /**
+   * Returns the reverse order in which the waypoints must be visited for an optimal journey.
+   */
+  public ArrayList<Point> getOrderedWaypoints() {
+    return orderedWaypoints;
```

Comment:
Be careful with storing sorted/ordered collections of things - you need to make sure that any modification to the source data causes this stored value to be updated too.

An option here is just to sort the underlying collection "on demand" here - i.e. sort it every time you need to return it. This may seem inefficient - if it turns out to be a bottleneck in your code then you might want to cache the value and do the bookkeeping required to make sure it is always updated, but for now my gut feeling is that this is going to be used infrequently enough that a minor extra millisecond to sort it "on demand" is not going to be a problem.

Reply:
I understand, I changed the logic so that the order of waypoints is computed "on demand".